### PR TITLE
Null-forgiving operator

### PR DESCRIPTION
**Pull request by Nigel-Ecma**
Null-forgiving operator

- Fix grammar
- Specify semantic restrictions
- Add to precedence table
- Add to MLR group

Addresses issue #1190

### CODE REVIEW

**Comment by Nigel-Ecma** [comment]
Note: The grammar validation will fail until the validator is updated. The validator is a work-in-progress and there is a mutual dependency between that and this PR. The PR for the validator will follow in the not too distant future…
@BillWagner – I suggest holding off merging this even once approved until a validator PR is good to go as well.

**Review by jskeet — APPROVED** [review]
Looks good to me, thanks Nigel.

File: standard/expressions.md
Comment (KalleOlaviNiemitalo) [inline standard/expressions.md]:
```suggestion
It is an error to apply the null-forgiving operator more than once to the same expression, intervening parentheses notwithstanding.
```

**Comment by KalleOlaviNiemitalo** [comment]
The wording does not seem to explain the effect of the null-forgiving operator in this case:

```csharp
class C {
    static void M() {
        object obj;
        N(out obj); // warning CS8600: Converting null literal or possible null value to non-nullable type.
        N(out obj!); // no warning
    }
    
    static void N(out object? o) { o = null; }
}
```

File: standard/expressions.md
Comment (KalleOlaviNiemitalo) [inline standard/expressions.md]: Was this restriction relaxed after C# 8?  Here's what I see at SharpLab:

```csharp
class C {
    static void M() {
        (string, string?) t1 = ("", null);
        (string?, string) t2;

        t2 = t1; // warning CS8619: Nullability of reference types in value of type '(string, string?)' doesn't match target type '(string?, string)'.
        t2 = t1!; // no warning, although (string, string?) is a value type
    }

    static void N() {
        (int, int) t1 = (0, 0);
        (int, int) t2;

        t2 = t1!; // no warning, although (int, int) is a value type
    }

    static void O() {
        int t1 = 0;
        int t2;

        t2 = t1!; // no warning, although int is a value type
    }
}
```

Comment (KalleOlaviNiemitalo) [reply standard/expressions.md]: Roslyn started allowing `e!` on value types in <https://github.com/dotnet/roslyn/pull/32090>, which is included in tag [Visual-Studio-2019-Version-16.0-Preview-2](https://github.com/dotnet/roslyn/releases/tag/Visual-Studio-2019-Version-16.0-Preview-2).  So Visual Studio 2019 version 16.0 already had this change in its C# 8 support.  [Visual Studio 2019 version 16.0 release notes for C#](https://learn.microsoft.com/en-us/visualstudio/releases/2019/release-notes-v16.0#c)

According to <https://github.com/dotnet/roslyn/issues/57142#issuecomment-947193118> on Oct 20, 2021, "the spec" was also changed to allow `!` on value types.  That change was perhaps <https://github.com/dotnet/csharplang/commit/e8ddd37721a7fbb5aff8a027f57d2cb25137b479> in `proposals/csharp-9.0/nullable-reference-types-specification.md`, making it look like `e!` on value types is only allowed starting from C# 9. But AFAIK no C# 8 compiler was released without this feature.

I suppose the committee still has the authority to omit that feature from the C# 8 standard, even if compilers support it.

File: standard/expressions.md
Comment (Nigel-Ecma) [reply standard/expressions.md]: Fixed, thanks @KalleOlaviNiemitalo

File: standard/expressions.md
Comment (BillWagner) [inline standard/expressions.md]: We should remove this paragraph. I checked with the team, and it was decided that `!` is allowed on a value type. The current compiler even allows it with C# 8 as the chosen version. I don't think even the C# 8 release had this change (as near as I can find in the history).

```suggestion
```

**Comment by Nigel-Ecma** [comment]
@KalleOlaviNiemitalo & @BillWagner – I only came here to fix the grammar! ;-)

Thanks Kalle for the research and further analysis.

It's clear the original spec was never in a released compiler and so we can go straight to allowing value types.

However as Kalle has shown the implemented operator does more than override (aka “forgive”) a compiler’s static analysis determination that a null might be present – it can be a “null*ability* forgiving operator” and cancel any nullability warnings that a compiler might choose to provide.

I’ve moved the PR to Draft while I determine a suitable spec.

**Comment by KalleOlaviNiemitalo** [comment]
Compiler behaviour for the null-forgiving operator in the left operand of an assignment expression looks complex:

```csharp
class C
{
    void M(
        ref object o,
        ref object p)
    {
        // error CS8598
        (o!) = null;

        // warning CS8600, despite ! operator
        (o!, p!) = (1, null);

        // no warning
        (o!, p!) = this;

        // no warning
        Deconstruct(out o!, out p!);
    }
    
    void Deconstruct(
        out object? o,
        out object? p)
    {
        o = null;
        p = null;
    }
}
```

**Comment by KalleOlaviNiemitalo** [comment]
Null-forgiving `!` would make perfect sense in the left operand of a compound assignment operator; but the compiler apparently doesn't allow that.  The new wording should make sure that the standard doesn't allow that either.

```csharp
public class C
{
    public static void M()
    {
        C? c = null;
        
        // error CS8598
        (c!) += new C();
        
        // no warning
        c = c! + new C();
    }
    
    public static C operator +(C c1, C c2)
    {
        return new C();
    }
}
```

**Comment by KalleOlaviNiemitalo** [comment]
warning CS8620 looks unjustified here; the type of the argument `(p, q)` should be `(object p, object? q)`, not `(object? p, object? q)`.  It seems the compiler makes the assignment to the tuple `(p, q)` change the type of `p` to nullable even before the assignment happens.  I suppose, even if it is a bug, it won't be a violation of the C# 8 standard, as compilers are allowed to output extra diagnostics.

```csharp
public class C
{
    public static void M()
    {
        object p = "";
        object? q = null;
        
        // warning CS8600: Converting null literal or possible null value to non-nullable type.
        // warning CS8620: Argument of type '(object? p, object? q)' cannot be used for parameter 't' of type '(object, object?)' in '(object?, object) C.operator +((object, object?) t, C c)' due to differences in the nullability of reference types.
        (p, q) = (p, q) + new C();     
    }
    
    public static (object?, object) operator +((object, object?) t, C c)
    {
        return (null, "");
    }
}
```

The same happens with a named method instead of an operator:

```csharp
public class C
{
    public static void M()
    {
        object p = "";
        object? q = null;
        
        // warning CS8600: Converting null literal or possible null value to non-nullable type.
        // warning CS8620: Argument of type '(object? p, object? q)' cannot be used for parameter 't' of type '(object, object?)' in '(object?, object) C.N((object, object?) t)' due to differences in the nullability of reference types.
        (p, q) = N((p, q));  
    }
    
    public static (object?, object) N((object, object?) t)
    {
        return (null, "");
    }
}
```

File: standard/expressions.md
Comment (Nigel-Ecma) [inline standard/expressions.md]: ### [Placeholder] 12.8.9

#### First Alternative

A certain compiler produces a compile time error in the following situation.

---
It is a compile-time error to apply the null-forgiving operator more than once to the same expression, intervening parentheses notwithstanding.

> *Example*: the following are all invalid:
>
> ```csharp
> var p = q!!;            // error: cannot apply the null_forgiving_operator more than once
> var s = ( ( m(t) ! ) )! // error: null_forgiving_operator applied twice to m(t)
> ```

---

#### Second Alternative

---
Remove placeholder

---

File: standard/expressions.md
Comment (Nigel-Ecma) [inline standard/expressions.md]: ### Placeholder 12.8.10.1

#### First Alternative

A certain compiler produces a compile time error in the following situation.

---
&lt;!-- [ToDo] C#9’s function pointers are also excluded, as the following restriction is stated in terms of what is included the text will probably be fine but this will need to be confirmed
-->
The *primary_expression* may be a *null_forgiving_expression* if and only if it has a *delegate_type*.

---

(The [ToDo] comment is part of the suggestion.)

#### Second Alternative

---
Remove placeholder

---

File: standard/expressions.md
Comment (Nigel-Ecma) [inline standard/expressions.md]: ### Placeholder 12.8.11

#### First Alternative

A certain compiler produces a compiler time error in the following situation.

---
&lt;!-- [ToDo] C#9’s function pointers are also excluded, as the following restriction is stated  in terms of what is included the text will probably be fine but this will need to be confirmed
-->
The optional *null_forgiving_operator* may be included if and only if the *null_conditional_member_access* or
*null_conditional_element_access* has a *delegate_type*.

---

(As before the [ToDo] comment is part of the suggestion.)

#### Second Alternative

---
Remove placeholder

---

File: standard/expressions.md
Comment (KalleOlaviNiemitalo) [inline standard/expressions.md]: Because there is 12.8.9.1 too, insert a "General" heading.

File: standard/expressions.md
Comment (Nigel-Ecma) [inline standard/expressions.md]: ### [Placeholder] 12.21.1

#### First Alternative

A certain compiler produces a compile time error in the following situation.

---
&lt;!-- [ToDo] This restriction also applies to null-coalescing assignment, ??= , which is a
     C# 8 feature not yet spec’ed – once it is need to ensure this restriction applies.
-->
The *unary_expression* left operand of an assignment may not be a *null_forgiving_expression*.

---

(As before the [ToDo] comment is part of the suggestion.)

#### Second Alternative

---
Remove placeholder

---

File: standard/unsafe-code.md
Comment (Nigel-Ecma) [inline standard/unsafe-code.md]: ### [Placeholder] 23.6.5

#### First Alternative

A certain compiler produces a compile time error in the following situation.

---
The *unary_expression* operand of an address-of operator may not be a *null_forgiving_expression*.

---

#### Second Alternative

---
Remove placeholder

---

File: standard/expressions.md
Comment (KalleOlaviNiemitalo) [inline standard/expressions.md]: "informs any ongoing analysis" apparently does not include propagating the information back to the variable from which the value originated.  [SharpLab](https://sharplab.io/#v2:CYLg1APgAgzABFATHAwnA3gWAFBzw+KAFjgFkAKASwDsAXAfjgEMBKDHfTuANyYCc4AIzgBeOFTosmAQgDcHLnl4CAxqPE1aU2XAD0uuNQD2cAO79qNAOYK8AX1txHsBCQBy5KAEYADI1bsuIo8/ELqnr5Sco6cynBqYhE+2noG5nyW1FaoAMoAHABsPj6ODth2QA===):

```csharp
public class C {
    public void M(int? a) {
        var b = (int)a!;
        var c = (int)a; // no warning
    }
    
    public void N(string? a) {
        var b = (string)a!;
        var c = (string)a; // warning CS8600
    }
}
```

In `N`, despite `a!` suggesting that `a` is known not to be null, the compiler issues a warning for the subsequent `(string)a` expression.

In `M`, this compiler does not warn about `(int)a`; however, this is apparently not caused by the null-forgiving operator in the preceding `(int)a!` expression, but rather by the compiler noticing that, if `a` were null, then `(int)a!` would throw and the subsequent `(int)a` would not be reached.

I did not check whether the specification already spells out that the null state analysis works this way.

**Review by Nigel-Ecma — COMMENTED** [review]
In previous discussions on this feature it was known to TG2 that there was one situation where a certain compiler produced a compile time error from a use of the null-forgiving operator, rather than the usual warning.

This seemingly diverged from the design goal that the null-forgiving operator only provides information to a compiler’s optional null state static analysis to effect the analysis and its warnings.

Those discussions led to the decision to include the requirement for the error in the Standard.

Since then a certain compiler’s source code has been reviewed and a total of 5 places where that compiler produces compile time errors from usage of the null-forgiving operator found – none of which appear to be specified in any of the feature design documents.

All 5 places, including the originally known one in §12.8.9, are now marked in this PR with placeholding HTML comments.

This review provides possible text for each placeholder if the particular error is to be included in the Standard and required of all implementations, or the choice of simply removing the placeholder.

Please debate away, either on all 5 errors as a whole or on the individual placeholder comments. TG2 will review the discussion before making final determinations.

File: standard/expressions.md
Comment (Nigel-Ecma) [reply standard/expressions.md]: I thought about this too but left it as due to the conditionally normative bit on line 1844 (“remainder of this subclause, including all of its subclauses”). This is probably one for @RexJaeschke to wordsmith.

File: standard/expressions.md
Comment (Nigel-Ecma) [reply standard/expressions.md]: @KalleOlaviNiemitalo – The specification leaves how the analysis works to the implementation, the null-forgiving operator “informs” the compiler that the user knows something but the compiler does not need to take heed… (well in `e!` the compiler is meant to change the determined null state, but remember that is for that specific `e`).

I’m not overly happy with the particular “informs any ongoing analysis” wording, and I tried and rejected other wordings, but it seems accurate – it only informs not, say, drives… Better wording suggestions welcome!

File: standard/types.md
Comment (Nigel-Ecma) [inline standard/types.md:730]: Some small wording corrections were made here; but between there being two distinct types, one just being the other with an *attribute*, the two being represented the same, and both capable of holding *exactly* the same set of values, it is less than clear… Two types? One and an attribute? Open invitation to have a go at wordsmithing!

File: standard/types.md
Comment (KalleOlaviNiemitalo) [inline standard/types.md]: _Moved to <https://github.com/dotnet/csharpstandard/pull/1178#discussion_r1836917012>_

The term "attribute" has a different meaning in [22 Attributes](https://github.com/dotnet/csharpstandard/blob/94c967e7c4cb49d9c4c1eec887b65793a2368c11/standard/attributes.md#22-attributes) so using it as part of *nullable_type_attribute* feels a bit unfortunate.  Not sure what would be a better term, though.  "Facet" as used in W3C XML Schema Definition Language (XSD) comes to mind but that doesn't feel right either.  Perhaps "specifier" as in *rank_specifier*?

Comment (Nigel-Ecma) [inline standard/types.md:723]: Note: *nullable_type_attribute* is not in the current grammar, it is defined in PR #1178

Comment (Nigel-Ecma) [reply standard/types.md]: @KalleOlaviNiemitalo – I forgot to add a note  yesterday that *nullabIe_type_attribute* is introduced in PR #1178; can you move your comment there?

**Comment by Nigel-Ecma** [comment]
General question: Anybody think an example explaining `null!` would be good and maybe also wish to suggest one, and where to include it, if so?

File: standard/types.md
Comment (BillWagner) [inline standard/types.md]: We've used "diagnostic messages" elsewhere to discuss the nullability analysis. I think we should be consistent.

File: standard/expressions.md
Comment (BillWagner) [inline standard/expressions.md]: This will produce the format we want in the final Word output without the markdown lint warnings:

```suggestion
<!-- markdownlint-disable MD028 -->

<!-- markdownlint-enable MD028 -->
```

File: standard/expressions.md
Comment (BillWagner) [inline standard/expressions.md:300]: Should the 2nd '-' also be the shorter character?

File: standard/expressions.md
Comment (BillWagner) [reply standard/expressions.md]: Here I'd prefer the first alternative. Not specifying that would limit later grammar.

File: standard/expressions.md
Comment (BillWagner) [inline standard/expressions.md]: I can't think of better wording, but I can think of weasel words:

```suggestion
message and may inform any ongoing analysis.
```

File: standard/expressions.md
Comment (BillWagner) [reply standard/expressions.md]: @Nigel-Ecma @KalleOlaviNiemitalo 

Here, I think the roslyn compiler assumes certain behavior of `Nullable<T>.operator T`, where it throws when the instance is null.

That is by design, to avoid false positives where possible.

**Review by BillWagner — COMMENTED** [review]
Overall this looks good. I had a few comments.

One other question for @jskeet:  Typically, we've used hard breaks only for paragraphs. I don't know if the hard returns will create problems for the word converter.

File: standard/expressions.md
Comment (Nigel-Ecma) [reply standard/expressions.md:300]: @BillWagner – No that one is an intentional en-dash (it will have been entered directly or as two dashes and smarten will have fixed it up).

File: standard/expressions.md
Comment (Nigel-Ecma) [reply standard/expressions.md]: @BillWagner mentions the known limitation argument.

Now I might be playing the devil’s advocate here, or maybe not 😉, but why does it not apply to *prefix* `!` or to operators such as `~`?

One could also ask why it doesn’t apply to negation as well, the following crazy code is valid (***please*** don't write it in real code!):

> ```
> int m = 0;
> int n = - - - --m;
> ```

which sets `m` to `-1` and `n` to `1`.

Given examples like this the apparent harmlessness of `x!!` stands out – it might earn a warning, but syntactic or semantic rules to prevent it?

Of course having the semantic rule as proposed here is also pretty harmless.

File: standard/expressions.md
Comment (Nigel-Ecma) [reply standard/expressions.md]: @BillWagner – I think that is better wording and I’ll do it unless others disagree.

File: standard/expressions.md
Comment (Nigel-Ecma) [reply standard/expressions.md]: @BillWagner @KalleOlaviNiemitalo – I would put it a bit more forcefully, “the Roslyn compiles *knows*” – and any compiler could. Whether a compiler analyses the bodies of called methods or uses the defined semantics of standard types etc. is part of the implementation specific latitude it has. Here's to the future VS vs. Rider null analysis competition! 🙂

File: standard/types.md
Comment (Nigel-Ecma) [reply standard/types.md]: @BillWagner – Thanks, I’ll go through and make sure they’re all the same.

Comment (Nigel-Ecma) [reply standard/types.md]: @BillWagner – A few moments later… The Standard actually uses diagnostic to refer to both errors and warnings, as it is key the null analysis produces warnings I’ll go with *diagnostic warnings* (which is already used) everywhere.

File: standard/expressions.md
Comment (KalleOlaviNiemitalo) [reply standard/expressions.md]: > why does it not apply to prefix ! or to operators such as ~?

1. Compatibility; cannot disallow `!!e` after it has been allowed in earlier versions of C# (and in C, where it has been used for ensuring that an `int` treated as a Boolean flag is either 0 or 1).
2. Prefix unary `!` and `-` can be overloaded so that `!!e` is not the same as `e` (not even the same type).

Those concerns don't apply to the null-forgiving operator because it's not used in pre-existing code and cannot be overloaded.

File: standard/expressions.md
Comment (KalleOlaviNiemitalo) [inline standard/expressions.md:1923]: OK, this also covers ref assignment.  [SharpLab](https://sharplab.io/#v2:D4AQzABCBMEMIQN4FgBQENUiALBAsgBQCUSamFUAjAAwD8EAzhALwQB2ArgDbcDc5ShgBOAUwBm1GhGGsZEpnwgB6ZRADuAQ2HsAluwDm8AMoAOAGxUAnIKGy2YyYyWqN2vYZMXrtyvflOAIQuauwA9m46+ga+AL6+vuBQeAByJGToQhCEYQBGAFaiAMYALnQANBAAJgCe7JoAtrpFdKQlcoRcvJVd3MQCmXYKhLX1TUWVeYWlpP6OECUhkR5GcGaWNoN+cvOLKmpaUZ5r3ptZMjsKJcH7HBGHK3Fo8ahAA=):

```csharp
public class C {
    public void M() {
        string? s = null;
        ref string r = ref s; // warning CS8619
        r = ref s; // warning CS8619
        r = ref s!; // no warning
    }
    
    public void N() {
        (object?, dynamic?) t = (null, null);
        ref (dynamic, object) r = ref t; // warning CS8619
        r = ref t; // warning CS8619
        r = ref t!; // no warning
    }
}
```

File: standard/expressions.md
Comment (jskeet) [inline standard/expressions.md:1754]:
```suggestion
A *null_conditional_member_access* consists of a *primary_expression* followed by the two tokens “`?`” and “`.`”, followed by an *identifier* with an optional *type_argument_list*, followed by zero or more *dependent_access*es any of which can be preceded by a *null_forgiving_operator*.
```

File: standard/expressions.md
Comment (jskeet) [reply standard/expressions.md]: +1 for the first alternative.

File: standard/expressions.md
Comment (jskeet) [reply standard/expressions.md]: I think the conditionally-normative aspect probably doesn't change our numbering requirements. Let's confirm on the call.

File: standard/expressions.md
Comment (jskeet) [reply standard/expressions.md]: I think I'd be okay with this being specified with the first alternative. Let's discuss in the call though.

File: standard/expressions.md
Comment (jskeet) [reply standard/expressions.md]: Just to check we're thinking about the same thing:

```csharp
ActionHolder? x = null;
x?.Action!(); // This is valid
x?.Dynamic!(); // This is not (CS8598)

record ActionHolder(Action? Action, dynamic? Dynamic);
```

Comment (jskeet) [reply standard/expressions.md]: Basically I suspect this should go hand-in-hand with "Placeholder 12.8.10.1" - if we go with the first alternative there, we should go with it here too, and vice versa.

File: standard/expressions.md
Comment (jskeet) [inline standard/expressions.md:3462]: But it's still a unary operator, right? If someone asked "what are the unary operators in C#" it should still be included. We may not want it in `unary_expression`, but I'm nervous about excluding it completely.

Let's talk about this one more.

File: standard/expressions.md
Comment (jskeet) [reply standard/expressions.md]: I don't have an opinion on this one yet - we need to talk about it.

File: standard/unsafe-code.md
Comment (jskeet) [reply standard/unsafe-code.md]: I'm fine with the first alternative here.

File: standard/expressions.md
Comment (Nigel-Ecma) [reply standard/expressions.md]: @jskeet – it is about the text of the conditionally normative bit, if “remainder of this subclause, including all of its subclauses” is *in* the first subclause then it needs to say something along the lines of the the remainder of this subclause, all its sibling subclauses, and any of the subclauses of these… and be readable! Hence why I suggested it is probably one for @RexJaeschke as our master wordsmith.

**Comment by Nigel-Ecma** [comment]
@BillWagner wrote:

> One other question for @jskeet: Typically, we've used hard breaks only for paragraphs. I don't know if the hard returns will create problems for the word converter.

The word convertor errors are not due to hard breaks. One of the subclause titles has emphasised text in it and the convertor doesn’t like that. It will be removed or replaced with quotes in the next commit.

File: standard/expressions.md
Comment (jskeet) [reply standard/expressions.md]: @Nigel-Ecma Ah right, thanks for clearing that up. Yes, that makes sense.

File: standard/unsafe-code.md
Comment (jnm2) [reply standard/unsafe-code.md]:
```cs
#nullable enable

unsafe
{
    int e = 4;
    _ = &e!; // Error
    _ = &(e!); // Error
    _ = (&e)!; // No error
}
```

Comment (jskeet) [reply standard/unsafe-code.md]: (Just for the record: I've been completely convinced by removing this, i.e. alternative 2)

File: standard/expressions.md
Comment (gafter) [inline standard/expressions.md:3460]: Call them unary expressions here?

File: standard/expressions.md
Comment (Nigel-Ecma) [reply standard/expressions.md:3462]: Topic moved to #1215

File: standard/expressions.md
Comment (Nigel-Ecma) [reply standard/expressions.md:1754]: Fixed

**Pull request merged**
